src/replayq_mem_ets_shared.erl

```diff
@@ -0,0 +1,168 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% @doc Shared ETS table for in-memory queue.
+%% Each item in the queue is a tuple of the form `{Id, Item}`.
+%% where `Id` is a tuple of `{self(), NextId}` and `NextId` is an increasing integer.
+%% Each owner process can only create one queue.
+-module(replayq_mem_ets_shared).
+
+-behaviour(replayq_mem).
+
+-export([
+    boot_init/0,
+    purge_by_owner/1
+]).
+
+-export([
+    new/1,
+    peek_all/1,
+    peek/1,
+    is_empty/1,
+    in/2,
+    in_r/2,
+    in_batch/2,
+    out/1,
+    destroy/1
+]).
+
+-export_type([queue/0]).
+
+%% parse-transform for ets:fun2ms/1
+-include_lib("stdlib/include/ms_transform.hrl").
+
+-define(ETS_TAB, ?MODULE).
+-define(KEY(PID, ID), {PID, ID}).
```

Comment:
If we do:

```suggestion
-define(KEY(PID, ID), {PID, {ID}}).
```

... then we can always find the first key with `ets:next(Tab, ?KEY(Owner, {}))`.